### PR TITLE
Consumer trainings endpoints for v2 endpoint instead of v1

### PR DESCRIPTION
**Pull request by kuceriklukas**

### CODE REVIEW

File: proto/asgt/v2/dataset_service.proto [hunk 1]
```diff
@@ -87,6 +87,8 @@ service DatasetService {
       get: "/v2/datasets/{dataset_name}/trainings/{number_of_trainings}"
     };
   }
+
+  rpc GetConsumerTrainings(ConsumerTrainingsRequest) returns (asgt.v2.type.ConsumerTrainingsResponse);
```
Comment (michalderdak) [inline proto/asgt/v2/dataset_service.proto:91]: I would call this `GetTrainings` and current `GetTrainings` --> `GetDatasetTrainings` 

furthermore, I'd add an HTTP1 path as well 

```
rpc GetTrainings (GetTrainingsRequest) returns (GetTrainingsResponse) {
    option (google.api.http) = {
      get: "/v2/trainings"
      body: "options" // not sure about this name
   };
}

message GetTrainingsRequest {
    ...
    TrainingsOptions options = xyz;
}

message TrainingsOptions {
    int limit = 1; // limit for no of trainings to be returned
}
```

Comment (michalderdak) [reply proto/asgt/v2/dataset_service.proto:91]: Also I think having response message in the `dataset_service.proto` is more fitting since request is here as well

File: proto/asgt/v2/type/trainings_response.proto [hunk 2]
```diff
@@ -14,7 +15,20 @@ message Training {
     google.protobuf.Timestamp finish_time = 5;
 }
 
+message ConsumerTraining {
+    asgt.type.Dataset dataset = 1;
```
Comment (michalderdak) [inline proto/asgt/v2/type/trainings_response.proto:19]: we can add dataset into `Training` message with id of `6` ... grpc is backwards compatible https://www.beautifulcode.co/blog/88-backward-and-forward-compatibility-protobuf-versioning-serialization

File: proto/asgt/v2/type/trainings_response.proto [hunk 3]
```diff
@@ -14,7 +15,20 @@ message Training {
...
+
 message TrainingsResponse {
     repeated Training trainings = 1;
 }
- 
\ No newline at end of file
+ 
+message ConsumerTrainingsResponse {
+    repeated ConsumerTraining trainings = 1;
+}
```
Comment (michalderdak) [inline proto/asgt/v2/type/trainings_response.proto:33]: I'd move this to `dataset_service.proto`

File: scripts/js_fixes.sh [hunk 4]
```diff
@@ -1,3 +1,3 @@
 export OUT_DIR=gen/js
 
-protoc -I./deps/googleapis deps/googleapis/google/type/* deps/googleapis/google/api/* -I./deps/protoc-gen-openapiv2 deps/protoc-gen-openapiv2/options/* --js_out=import_style=commonjs:$OUT_DIR
+protoc -I./deps/googleapis deps/googleapis/google/type/* deps/googleapis/google/api/* -I./deps/protoc-gen-openapiv2 deps/protoc-gen-openapiv2/options/* -I./deps/validation deps/validation/validate/* --js_out=import_style=commonjs:$OUT_DIR
```
Comment (michalderdak) [inline scripts/js_fixes.sh:3]: I'm still kinda puzzled why is this needed. I get that imports are not working without all of its dependencies but I still think server-side validation shouldn't require to have validation import for client-side code

**Review by michalderdak — CHANGES_REQUESTED** [review]

File: proto/asgt/v2/dataset_service.proto (hunk 1 above)
Comment (kuceriklukas) [reply proto/asgt/v2/dataset_service.proto:91]: I absolutely agree. Just regarding the code you put here, since `GetTrainings` should be a get request, it doesn't really have a post body, so I'd suggest doing something similar to what Miro implemented in his previous trainings call, something along those lines:
```proto
  rpc GetDatasetTrainings (GetDatasetTrainingsRequest) returns (TrainingsResponse) {
    option (google.api.http) = {
      get: "/v2/datasets/{dataset_name}/trainings/{number_of_trainings}"
    };
  }

  rpc GetTrainings(GetTrainingsRequest) returns (TrainingsResponse) {
    option (google.api.http) = {
      get: "/v2/trainings/{number_of_trainings}"
    };
  }

  ...

  message GetTrainingsRequest {
    uint32 number_of_trainings = 1;
  }

  message TrainingsResponse {
    repeated asgt.v2.type.Training trainings = 1;
  }
```

Comment (michalderdak) [reply proto/asgt/v2/dataset_service.proto:91]: the body would map to query parameters... there's an issue with miro's implementation where `get: "/v2/trainings/{number_of_trainings}"` number of trainings is a place for an ID of training and not a limit ... 

see https://github.com/googleapis/googleapis/blob/master/google/api/http.proto#L128

Comment (michalderdak) [reply proto/asgt/v2/dataset_service.proto:91]: instead of `"/v2/trainings/{number_of_trainings}"` we want `"/v2/trainings?limit=10"` because then if needed we can do `"/v2/trainings/{training_id}"` to return specific trainings

Comment (kuceriklukas) [reply proto/asgt/v2/dataset_service.proto:91]: Alright, it makes sense that the request structure should be `/v2/trainings/{training_id}`, in this case `/v2/trainings?limit=10` - I haven't realized this, my bad.

However, after a little bit of digging, I don't think you're right about the `body` being mapped to the query parameters. It doesn't say anywhere actually, nor I could find implementation of `get` with a `body`. What seems like a solution (https://github.com/googleapis/googleapis/blob/0c2e4540e01637d1964d141e68e8cd20619e4c6e/google/api/http.proto#L89) is something like this:
```proto
  rpc GetDatasetTrainings (GetDatasetTrainingsRequest) returns (TrainingsResponse) {
    option (google.api.http) = {
      get: "/v2/datasets/{dataset_name}/trainings"
    };
  }

  rpc GetTrainings(GetTrainingsRequest) returns (TrainingsResponse) {
    option (google.api.http) = {
      get: "/v2/trainings"
    };
  }

  ...

  message GetDatasetTrainingsRequest {
    string dataset_name = 1 [...validation rules];  // maps to the url placeholder `dataset_name`
    int64 limit = 2;    // according to the docs, this should translate into URL query parameter `limit`.
}

  message GetTrainingsRequest {
    int64 limit = 1;   // according to the docs, this should translate into URL query parameter `limit`.
  }
```

The `body` seems to be reserved only for the http requests that actually can contain http body (at least according to this https://github.com/googleapis/googleapis/blob/0c2e4540e01637d1964d141e68e8cd20619e4c6e/google/api/http.proto#L124)

File: scripts/js_fixes.sh (hunk 4 above)
Comment (kuceriklukas) [reply scripts/js_fixes.sh:3]: For future reference - this seems to be a general problem with the validation library. For now, one can only decide to either include the validation library in the javascript generated files or do some text replacing on the generated files afterwards. See [this stackoverflow post](https://stackoverflow.com/questions/62983718/how-to-exclude-an-import-on-client-side) for example.

File: proto/asgt/v2/dataset_service.proto [hunk 5]
```diff
@@ -126,7 +132,15 @@ message TruncateDatasetRequest {
   string name = 1 [(validate.rules).string = { pattern: "^[A-Za-z0-9.][A-Za-z0-9_.>-]*$", max_bytes: 256 }]; // text-no-spaces
 }
 
-message TrainingsRequest {
+message GetDatasetTrainingsRequest {
   string dataset_name = 1 [(validate.rules).string = { pattern: "^[A-Za-z0-9.][A-Za-z0-9_.>-]*$", max_bytes: 256 }]; // text-no-spaces
-  uint32 number_of_trainings = 2;
+  int64 limit = 2; // mapped to URL query parameter `limit`
```
Comment (michalderdak) [inline proto/asgt/v2/dataset_service.proto:137]: I think I'd like this more to be an "option" message containing the limit... e.g. 

```
message GetDatasetTrainingsRequest {
  message TrainingRequestOptions {
    int64 limit = 1;
  }
  string dataset_name = 1 [val];
  message TrainingRequestOptions options = 2;
```

Comment (michalderdak) [reply proto/asgt/v2/dataset_service.proto:137]: No specific reason behind it.. just saw it being done that way

**Review by michalderdak — APPROVED** [review]

File: proto/asgt/v2/dataset_service.proto (hunk 5 above)
Comment (kuceriklukas) [reply proto/asgt/v2/dataset_service.proto:137]: Aaah I get now what you meant with the `options`. Yes, it would particularly make sense if we would have more calls that need the same options - so maybe not now, but in the future it could be useful. I see if I can do it before merging this. I'm just not sure how the URL query parameters will like the nested messages... but it should be possible.

Comment (michalderdak) [reply proto/asgt/v2/dataset_service.proto:137]: well "maybe in the future" isn't backwards compatible 😅 

Also IDK how all of this works with query params as this is completely new to me

**Review by michalderdak — APPROVED** [review]
